host/crypto/bcrypt/cert.c

```diff
@@ -0,0 +1,1295 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+#include <assert.h>
+#include <openenclave/bits/result.h>
+#include <openenclave/bits/safecrt.h>
+#include <openenclave/bits/safemath.h>
+#include <openenclave/internal/cert.h>
+#include <openenclave/internal/pem.h>
+#include <openenclave/internal/raise.h>
+#include <openenclave/internal/utils.h>
+
+#include "../magic.h"
+#include "bcrypt.h"
+#include "crl.h"
+#include "ec.h"
+#include "key.h"
+#include "pem.h"
+#include "rsa.h"
+
+/*
+**==============================================================================
+**
+** Local definitions:
+**
+**==============================================================================
+*/
+
+#define _OE_CERT_CHAIN_LENGTH_ANY 0
+
+static const DWORD _OE_DEFAULT_GET_CRL_FLAGS = CERT_STORE_SIGNATURE_FLAG |
+                                               CERT_STORE_TIME_VALIDITY_FLAG |
+                                               CERT_STORE_BASE_CRL_FLAG;
+
+static const CERT_CHAIN_POLICY_PARA _OE_DEFAULT_CERT_CHAIN_POLICY = {
+    .cbSize = sizeof(CERT_CHAIN_POLICY_PARA),
+    .dwFlags = 0};
+
+static const CERT_STRONG_SIGN_PARA _OE_DEFAULT_SIGN_PARAMS = {
+    .cbSize = sizeof(CERT_STRONG_SIGN_PARA),
+    .dwInfoChoice = CERT_STRONG_SIGN_OID_INFO_CHOICE,
+    .pszOID = szOID_CERT_STRONG_SIGN_OS_1};
+
+static const CERT_CHAIN_PARA _OE_DEFAULT_CERT_CHAIN_PARAMS = {
+    .cbSize = sizeof(CERT_CHAIN_PARA),
+    .RequestedUsage = {{0}},
+    .RequestedIssuancePolicy = {{0}},
+    .dwUrlRetrievalTimeout = 0,
+    .fCheckRevocationFreshnessTime = FALSE,
+    .dwRevocationFreshnessTime = 0,
+    .pftCacheResync = NULL,
+    .pStrongSignPara = &_OE_DEFAULT_SIGN_PARAMS,
+    .dwStrongSignFlags = 0};
+
+static const DWORD _OE_DEFAULT_CERT_CHAIN_FLAGS =
+    CERT_CHAIN_CACHE_END_CERT | CERT_CHAIN_REVOCATION_CHECK_CACHE_ONLY |
+    CERT_CHAIN_CACHE_ONLY_URL_RETRIEVAL | CERT_CHAIN_REVOCATION_CHECK_CHAIN;
+
+typedef struct _cert
+{
+    uint64_t magic;
+    PCCERT_CONTEXT cert;
+} cert_t;
```

Comment:
Elsewhere we assert that the sizeof(impl type) <= sizeof(container type). Is there some assertion we could add here?

Reply:
Added checks for `cert_t` and `cert_chain_t` both in bcrypt and openssl implementations. Also renamed `Cert` and `CertChain` names in openssl impl to conform to snake casing.